### PR TITLE
fedora_bot: limit the number of dist-git retries

### PR DESCRIPTION
**Pull request by ondrejbudai**
The previous implementation of dist-git HTTP call retries is very naïve.
There's no retry count limit and we retry even on status codes like 4XX that
usually aren't retryable.

What's even worse is that if we are stuck in an infinite retrying loop,
github actions will kill the script after 6 hours without showing any logs
so we don't even know what went wrong.

This commit overhauls the retrying logic:
- maximum of 3 retries
- backoff factor was added
- retries only on retryable 5XX errors

This is implemented using the handy requests.adapters.Retry adapter,
see https://urllib3.readthedocs.io/en/stable/reference/urllib3.util.html

### CODE REVIEW

File: fedora_bot.py [hunk 1]
```diff
@@ -101,24 +102,37 @@ def slack_notify(message: str):
...
+    Returns a configured http client for Fedora's dist-git. We need a custom
+    client in order to implement retries. These are needed because
+    src.fedoraproject.org proved to be slightly unreliable and without
+    retries, some fedora-bot runs failed.
+    """
+    retry_adapter = Retry(
+        total=3,
+        backoff_factor=0.1,
+        # POST isn't generally idempotent but we only use it to merge PRs
+        # which actually idempotent is.
```
Comment (martinpitt) [inline fedora_bot.py:116]: The grammar force use you must! :wink:

File: fedora_bot.py [hunk 2]
```diff
@@ -101,24 +102,37 @@ def slack_notify(message: str):
...
+        allowed_methods=["POST"] + list(Retry.DEFAULT_ALLOWED_METHODS),
+        # Enable retrying on retryable http codes.
+        status_forcelist=[500, 502, 502, 503, 504],
```
Comment (martinpitt) [inline fedora_bot.py:119]: Duplicate 502 -- this looks like you meant 501?

**Review by martinpitt — CHANGES_REQUESTED** [review]
This is quite a nice approach! And admittedly the first-ever feature of requests that is worth using a separate dependency for (depending on the use case).

File: fedora_bot.py (hunk 1 above)
Comment (ondrejbudai) [reply fedora_bot.py]: I blame my native language! :sweat_smile: (I hope I got it correctly this time).

File: fedora_bot.py (hunk 2 above)
Comment (ondrejbudai) [reply fedora_bot.py]: Nah, 501 is Not Implemented, this is not retryable. I want:

- 500 Internal Server Error
- 502 Bad Gateway
- 503 Service Unavailable
- 504 Gateway timeout

**Comment by ondrejbudai** [comment]
Force-pushed to fix @martinpitt's suggestions.

**Review by martinpitt — APPROVED** [review]
Thanks! LGTM.

**Pull request merged**